tests/test_unit_conversion.py

```diff
@@ -1,80 +1,1235 @@
+
+        assert equation[1].lhs == symbol_b
+        with pytest.raises(units.UnitError):
+            # cellml file states b (per_ms) = power(a (ms), 1 (second))
+            bad_units_model.units.summarise_units(equation[1].rhs)
+
+    def test_add_input_state_variable(self, local_model):
+        """ Tests the Model.add_input function that changes units.
+        This particular test is when a state variable is being converted
+
+        For example::
+
+            Original model
+                var{time} time: ms {pub: in};
+                var{sv11} sv1: mV {init: 2};
+
+                ode(sv1, time) = 1{mV_per_ms};
+
+        convert_variable(sv1, volt, DataDirectionFlow.INPUT)
+
+            creates model
+                var{time} time: ms {pub: in};
+                var{sv11} sv1_converted: V {init: 0.002};
+                var sv1 mV {init: 2}
+                var sv1_orig_deriv mV_per_ms
+
+                sv1 = 1000 * sv1_converted
+                sv1_orig_deriv = 1{mV_per_ms}
+                ode(sv1_converted, time) = 0.001 * sv1_orig_deriv
+        """
+        # original state
+        def test_original_state(local_model):
+            assert len(local_model.variables()) == 3
+            symbol_a = local_model.get_symbol_by_cmeta_id('sv11')
+            symbol_t = local_model.get_symbol_by_cmeta_id('time')
+            assert local_model.get_initial_value(symbol_a) == 2.0
+            assert symbol_a.units == 'mV'
+            assert symbol_t.units == 'ms'
+            assert len(local_model.equations) == 1
+            assert str(local_model.equations[0]) == 'Eq(Derivative(_env_ode$sv1, _environment$time), _1.0)'
+            state_symbols = local_model.get_state_symbols()
+            assert len(state_symbols) == 1
+            assert symbol_a in state_symbols
+            return True
+
+        mV_unit = local_model.get_units('mV')
+        volt_unit = local_model.get_units('volt')
+        original_var = local_model.get_symbol_by_name('env_ode$sv1')
+
+        assert test_original_state(local_model)
+        # test no change in units
+        newvar = local_model.convert_variable(original_var, mV_unit, DataDirectionFlow.INPUT)
+        assert newvar == original_var
+        assert test_original_state(local_model)
+
+        # change mV to V
+        newvar = local_model.convert_variable(original_var, volt_unit, DataDirectionFlow.INPUT)
+        assert newvar != original_var
+        assert len(local_model.variables()) == 5
+        symbol_a = local_model.get_symbol_by_cmeta_id('sv11')
+        assert local_model.get_initial_value(symbol_a) == 0.002
+        assert symbol_a.units == 'volt'
+        assert symbol_a.name == 'env_ode$sv1_converted'
+        symbol_t = local_model.get_symbol_by_cmeta_id('time')
+        assert symbol_t.units == 'ms'
+        symbol_orig = local_model.get_symbol_by_name('env_ode$sv1')
+        assert symbol_orig.units == 'mV'
+        assert local_model.get_initial_value(symbol_orig) == 2.0
```

Comment:
Hmm, we ought to delete the initial_value from the original variable, because it's now defined by an equation instead. Sorry I didn't catch that in the code above!

Reply:
unless its an output (or unless there are other problems in my code !)

Reply:
Indeed!